File: C/ds/StackUsingQueue.c
/* C Program to implement a Stack using Queue.Here we will try to realise STACK
    using the properties OR functions of QUEUE and by making pop operation
   costly*/

#include <malloc.h>
#include <stdio.h>

struct Node {
  int data;
  struct Node *next;
};

/*Maintaining a front,rear pointer
 where insertion takes place at rear end and deletion at front
 and O(1) time is taken  for insertion by maintaining a rear pointer
*/
struct Node *front = NULL, *rear = NULL;
// Store size of queue
int qsize = 0;

// check whether queue is empty or not
int isEmpty(struct Node *q) {
  if (q == NULL) {
    // return true i.e Queue is empty
    return 1;
  }
  return 0;
}

// Enqueue i.e insert at the REAR end
int enQueue(int item) {
  struct Node *q;
  // creation of new node
  q = (struct Node *)malloc(sizeof(struct Node));
  q->data = item;
  q->next = NULL;

  /* If queue is empty
  i.e front is pointing to NULL update front*/
  if (front == NULL) {
    front = q;
  } else {
    (rear)->next = q;
  }

  rear = q;
  // increase current queue size
  qsize++;
  return 0;
}

// Dequeue i.e delete at the FRONT end
int deQueue() {
  int delElement;
  /*Queue is empty
  dequeue not possible*/
  if (isEmpty(front)) {
    return -1;
  } else {
    // store the dequeued element
    delElement = front->data;
    (front) = (front)->next;

    if ((front) == NULL) {
      (rear) == NULL;
    }
    // reduce the current queue size
    qsize--;
  }
  return delElement;
}

void push(int data) { enQueue(data); }

int pop() {
  int i, element;
  /* dequeue upto n-1 elements in the queue
     and enqueue the respective dequeued element
         at the REAR end of the queue*/
  for (i = 0; i < qsize - 1; i++) {
    element = deQueue();
    enQueue(element);
  }
  /* dequeue the element which will be first value of the STACK
   and return the element*/
  element = deQueue();
  if (element == -1) {
    printf("\n Stack is empty");
    return 0;
  }

  return element;
}

void display(struct Node *temp) {

  while (temp != NULL) {
    printf("%d ", temp->data);
    temp = temp->next;
  }
}

int main() {

  int i, val, size;
  printf("Enter the size: ");
  scanf("%d", &size);

  printf("Enter the numbers:\n");
  for (i = 0; i < size; i++) {
    scanf("%d", &val);
    push(val);
  }

  printf("Stack size:%d\n", qsize);
  printf("Popped Elements are:\n");
  while (qsize > 0) {
    printf("\n%d ", pop());
  }
  printf("\nStack size:%d", qsize);
  return 0;
}

/*
  Sample input/output:
  Enter the size: 5
  Enter the numbers:
  1
  2
  3
  4
  5
  Stack Size:5
  Popped Elements are:
  5
  4
  3
  2
  1
  Stack Size:0

  Time complexity: O(1) for push, O(n) for pop
  Space complexity: O(n)
 */

File: Java/ds/StackUsingQueue.java
/* Java Program to implement a stack using two queues
by making the push operation costly*/

import java.util.*;

public class StackUsingQueue {
    static class Stack {
        // the 2 queues
        Queue<Integer> qu1 = new LinkedList<>();
        Queue<Integer> qu2 = new LinkedList<>();

        // storing current size
        int currentSize;

        Stack() {
            currentSize = 0;
        }

        void push(int num) {
            currentSize++;

            // first, push num into an empty qu2
            qu2.add(num);

            // now we push the other elements from qu1 to qu2
            while (!qu1.isEmpty()) {
                qu2.add(qu1.peek());
                qu1.remove();
            }
        }

        void pop() {
            if (qu2.isEmpty()) {
                return;
            }

            // remove element and reduce size if qu2 not empty
            qu2.remove();
            currentSize--;
        }

        int top() {
            if (qu2.isEmpty()) {
                return -1;
            }

            // return the top element
            return qu2.peek();
        }

        int size() {
            return currentSize;
        }
    }

    public static void main(String[] args) {
        Stack stack = new Stack();

        Scanner sc = new Scanner(System.in);
        System.out.println("Enter size:");
        int size = sc.nextInt();
        System.out.println("Enter the numbers:");
        for (int i = 0; i < size; i++) {
            stack.push(sc.nextInt());
        }
        sc.close();

        System.out.println("Current Size:" + stack.size());
        for (int i = size; i > 0; i--) {
            System.out.println(stack.top());
            stack.pop();
        }
        System.out.println("Current Size:" + stack.size());
    }
}

/**
 * Sample input/output:
 * 
 * Enter size:
 * 5
 * Enter the numbers:
 * 5
 * 4
 * 3
 * 2
 * 1
 * Current Size:5
 * 5
 * 4
 * 3
 * 2
 * 1
 * Current Size:0
 * 
 * Time complexity: O(n) for push, O(1) for pop
 * Space complexity: O(n)
 */
